### PR TITLE
OCPBUGS-74937: “Import from Git” should not force HTTPS/443 for Git providers (GitHub/GitLab/Bitbucket)

### PR DESCRIPTION
**Pull request by krishagarwal278**
## Release Notes

* **Bug Fixes**
  * Improved handling of custom Git service URLs with different schemes (HTTP/HTTPS) and non-default ports for Bitbucket Server, GitHub Enterprise, and GitLab instances.

* **Tests**
  * Added test coverage verifying correct URL scheme and port preservation across Git service integrations.

### CODE REVIEW

**Comment by yanpzhan** [comment]
Checked on cluster launched against the pr.
Clicked 'Import from Git' from '+' drowdown list in masthead, opened developer tool, clicked 'Network' tab, 
1. In Git Repo URL in put 'https://gh.local:8443/org/repo', chose 'GitHub' Git Type. Filtered 'local' in 'Network' tab.
Could see the api request is using https schema and port 'https://gh.local:8443/api/v3/repos/org/repo'.
2. In Git Repo URL in put 'http://gh.local:8080/org/repo', chose 'GitHub' Git Type. Filtered 'local' in 'Network' tab.
Could see the api request is using http schema and port 'http://gh.local:8080/api/v3/repos/org/repo'.
3. In Git Repo URL in put 'http://gitlab.local:8080/grp/repo', chose 'GitHub' Git Type. Filtered 'local' in 'Network' tab.
Could see the api request is using http schema and port 'http://gitlab.local:8080/api/v3/repos/grp/repo'
4. In Git Repo URL in put 'http://bb.local:7990/scm/prj/repo', chose 'GitHub' Git Type. Filtered 'local' in 'Network' tab.
Could see the api request is using http schema and port 'http://bb.local:7990/rest/api/1.0/projects/scm/repos/prj'
5. In Git Repo URL in put 'http://gitea.local:3000/user/repo'. Filtered 'local' in 'Network' tab.
Could see the api request is using https schema without port: 'https://gitea.local/api/v1/repos/user/repo'

@krishagarwal278 Your fix works for github/gitlab/bitbucket(in test result 1 to 4). But seems there is not fix for gitea type(in test result 5), need gitea type be fixed?

**Comment by krishagarwal278** [comment]
hello @yanpzhan thank you for verifying this fix! 
gitea type(in test result 5) Is being fixed in [https://github.com/openshift/console/pull/15975](https://github.com/openshift/console/pull/15975)

**Comment by krishagarwal278** [comment]
@cajieh @jhadvig can I get the approved/lgtm labels for this one please, thank you!
/assign @jhadvig

File: frontend/packages/git-service/src/services/github-service.ts [hunk 1]
```diff
@@ -61,12 +62,14 @@ export class GithubService extends BaseService {
...
-    const { name, owner, source } = GitUrlParse(this.gitsource.url);
+    const { name, owner, protocols, port, resource } = GitUrlParse(this.gitsource.url);
+    const protocol = protocols?.[0] || 'https';
+    const host = port ? `${protocol}://${resource}:${port}` : `${protocol}://${resource}`;
     const contextDir = this.gitsource.contextDir?.replace(/\/$/, '') || '';
     return {
       repoName: name,
       owner,
-      host: source,
+      host,
```
[the remark this answers is not in the document]
Comment (rhamilto) [reply frontend/packages/git-service/src/services/github-service.ts:76]: Agreed

[the remark this answers is not in the document]
Comment (krishagarwal278) [reply frontend/packages/git-service/src/services/github-service.ts:76]: @coderabbitai @rhamilto This has been fixed in commit d6d31bb188fe9031dea10675b122aa24d165de91
Now both (gitlab and github) validate protocols to only accept http/https, defaulting to https for other schemes like SSH, preventing malformed API URLs

**Comment by rhamilto** [comment]
Claude has additional feedback that is worth examining:

# Pull Request Review: OCPBUGS-74937

**PR**: https://github.com/openshift/console/pull/15978
**Title**: "Import from Git" should not force HTTPS/443 for Git providers (GitHub/GitLab/Bitbucket)
**Author**: krishagarwal278
**Date**: 2026-02-04

## Summary

This PR fixes a bug where "Import from Git" forces HTTPS/443 for custom Git provider instances, breaking support for HTTP-only servers or non-standard ports.

**Changes**:
- 6 files changed
- +73 additions, -11 deletions
- Affects: GitHub, GitLab, and Bitbucket service implementations
- Adds test coverage for URL scheme and port preservation

## Review Findings

### ✅ Positives

1. **Addresses a real bug**: Custom Git providers with HTTP or non-standard ports are currently broken
2. **Good test coverage**: Each modified service has new test cases
3. **Backwards compatible**: Defaults to HTTPS when protocol cannot be determined

### ⚠️ Issues Found

#### 1. CRITICAL: Missing Gitea Service

The PR updates GitHub, GitLab, and Bitbucket services but **misses Gitea Service** (`gitea-service.ts:67`), which has the same bug:

```typescript
const host = `https://${resource}`;  // Also hardcodes HTTPS
```

**Impact**: Gitea users with HTTP or custom ports will still experience the same bug.

**Recommendation**: Apply the same fix to `gitea-service.ts` for consistency.

---

#### 2. Inconsistent Implementation Across Services

Three different approaches are used:

**GitHub/GitLab**: Extract protocol from `GitUrlParse`:
```typescript
const { protocols, port, resource } = GitUrlParse(this.gitsource.url);
const rawProtocol = protocols?.[0];
const protocol = isHttpProtocol ? rawProtocol : 'https';
```

**Bitbucket**: Uses `new URL()` with try-catch:
```typescript
const url = new URL(this.gitsource.url);
if (url.protocol === 'http:' || url.protocol === 'https:') {
  protocol = url.protocol.replace(':', '');
}
```

**Recommendation**: Standardize on `GitUrlParse` since it's already used throughout the codebase and handles various Git URL formats better (SSH, HTTPS, etc.).

---

#### 3. Bitbucket Inconsistency

Bitbucket only modifies `baseURL` in the constructor but doesn't update `getRepoMetadata()`, so `this.metadata.host` remains just a hostname while in GitHub/GitLab it's a full URL. This creates an inconsistency.

**Current state**:
- GitHub/GitLab: `metadata.host` = `"http://example.com:3000"` (full URL)
- Bitbucket: `metadata.host` = `"example.com"` (hostname only)
- Bitbucket: `baseURL` = `"http://bb.example.com:7990/rest/api/1.0"` (full URL with path)

The `baseURL` is used directly in API calls (bitbucket-service.ts:106, 134, etc.), so this works functionally, but it's inconsistent with the other services' approach.

**Recommendation**: Update Bitbucket's `getRepoMetadata()` to match the GitHub/GitLab pattern for consistency.

---

#### 4. Semantic Change to `RepoMetadata.host`

The `host` field changes from hostname to full URL (with protocol and port). This is a **breaking change** if any external code depends on this field.

**Before**: `"github.com"`
**After**: `"https://github.com"` or `"http://example.com:3000"`

**Considerations**:
- Are there any consumers of `RepoMetadata.host` outside these service files?
- Plugin API implications?
- Webhook creation code may be affected

**Recommendations**:
1. Search codebase for other uses of `metadata.host`
2. Consider renaming to `baseUrl` or `apiUrl` for clarity
3. Or keep `host` as hostname and add separate `protocol`/`port` fields
4. Document this breaking change in the PR description

---

#### 5. Port Handling Logic Issue

**github-service.ts:68-70**
```typescript
const host = isHttpProtocol && port
  ? `${protocol}://${resource}:${port}`
  : `${protocol}://${resource}`;
```

This only includes the port if the protocol is HTTP/HTTPS **and** a port exists. This means:
- ✅ `http://example.com:3000` → includes port
- ❌ `https://example.com:8443` → includes port only if detected as HTTPS
- ⚠️ Non-HTTP protocols default to HTTPS but lose port information

**Issue**: If `isHttpProtocol` is `false` (e.g., SSH URL), the code defaults to `https://` but drops the port.

**Suggested improvement**:
```typescript
const port = GitUrlParse(this.gitsource.url).port;
const host = port
  ? `${protocol}://${resource}:${port}`
  : `${protocol}://${resource}`;
```

This ensures ports are preserved regardless of the original protocol.

---

#### 6. Test Coverage Gaps

The tests cover HTTP with custom ports:
- ✅ `http://example.com:3000`
- ✅ `http://bb.example.com:7990`

**Missing test cases**:
- HTTPS with custom ports (e.g., `https://example.com:8443`)
- Standard ports - should `:80` for HTTP and `:443` for HTTPS be included or omitted?
- Edge cases like malformed URLs
- Git SSH URLs (e.g., `git@example.com:org/repo.git`)

---

#### 7. Misleading Comment in Bitbucket Service

**bitbucket-service.ts:44-51**
```typescript
let protocol = 'https';
try {
  const url = new URL(this.gitsource.url);
  if (url.protocol === 'http:' || url.protocol === 'https:') {
    protocol = url.protocol.replace(':', '');
  }
} catch (e) {
  // fall through to git-url-parse based handling
}
```

The comment says "fall through to git-url-parse based handling" but nothing happens in the catch block - the code just uses the default 'https'. The comment is misleading.

**Recommendation**: Update comment to:
```typescript
} catch (e) {
  // Invalid URL format, default to https
}
```

---

## Specific Code Review

### github-service.ts

**Lines 47-49** - Constructor baseUrl setup:
```typescript
const { resource, port } = GitUrlParse(this.gitsource.url);
const isGithubDotCom = resource === 'github.com' && !port;
const baseUrl = isGithubDotCom ? null : `${this.metadata.host}/api/v3`;
```
✅ Correct - relies on `metadata.host` already containing the full URL

**Lines 64-75** - getRepoMetadata():
```typescript
const { protocols, port, resource } = GitUrlParse(this.gitsource.url);
const rawProtocol = protocols?.[0];
const isHttpProtocol = rawProtocol === 'http' || rawProtocol === 'https';
const protocol = isHttpProtocol ? rawProtocol : 'https';
const host = isHttpProtocol && port
  ? `${protocol}://${resource}:${port}`
  : `${protocol}://${resource}`;
```
⚠️ Should include port even for HTTPS if non-standard (see Issue #5 above)

**Lines 184-188** - createRepoWebhook():
```typescript
const { resource, port } = GitUrlParse(this.gitsource.url);
const isGithubDotCom = resource === 'github.com' && !port;
const AddWebhookBaseURL = isGithubDotCom
  ? `https://api.github.com`
  : `${this.metadata.host}/api/v3`;
```
✅ Looks good

### gitlab-service.ts

**Lines 80-92** - getRepoMetadata():
```typescript
const { name, owner, protocols, port, resource, full_name: fullName } = GitUrlParse(
  this.gitsource.url,
);
const contextDir = removeLeadingSlash(this.gitsource.contextDir);
const rawProtocol = protocols?.[0];
const isHttpProtocol = rawProtocol === 'http' || rawProtocol === 'https';
const protocol = isHttpProtocol ? rawProtocol : 'https';
const host = isHttpProtocol && port
  ? `${protocol}://${resource}:${port}`
  : `${protocol}://${resource}`;
```
⚠️ Same port handling issue as GitHub service

### bitbucket-service.ts

**Lines 44-47** - Constructor:
```typescript
if (this.metadata.host !== 'bitbucket.org') {
  let protocol = 'https';
  try {
    const url = new URL(this.gitsource.url);
```
⚠️ Different approach than GitHub/GitLab - should be standardized

**Lines 90-96** - getRepoMetadata():
```typescript
const { name, owner, host } = ParseBitbucketUrl(this.gitsource.url);
const contextDir = this.gitsource.contextDir?.replace(/\/$/, '') || '';
return {
  repoName: name,
  owner,
  host,
```
⚠️ Returns hostname only, not full URL like GitHub/GitLab

---

## Recommendations

### Priority 1: Critical

1. **Add Gitea Service** (`gitea-service.ts`) to the PR for completeness
2. **Fix port handling** - include non-standard ports regardless of protocol

### Priority 2: High

3. **Standardize implementation** - use `GitUrlParse` consistently across all services
4. **Update Bitbucket** to modify `getRepoMetadata()` for consistency with GitHub/GitLab

### Priority 3: Medium

5. **Add test cases** for:
   - HTTPS with custom ports
   - Edge cases (malformed URLs, standard ports)
6. **Fix misleading comment** in Bitbucket service catch block
7. **Document the breaking change** to `RepoMetadata.host` in the PR description
8. **Search for other consumers** of `metadata.host` to assess impact

### Priority 4: Nice to Have

9. Consider semantic changes to `RepoMetadata` interface for clarity (`baseUrl` vs `host`)
10. Add integration tests if possible

---

## Suggested Implementation for Port Handling

Replace the port handling logic in GitHub and GitLab services:

**Current (problematic)**:
```typescript
const host = isHttpProtocol && port
  ? `${protocol}://${resource}:${port}`
  : `${protocol}://${resource}`;
```

**Suggested**:
```typescript
const { protocols, port, resource } = GitUrlParse(this.gitsource.url);
const rawProtocol = protocols?.[0];
const isHttpProtocol = rawProtocol === 'http' || rawProtocol === 'https';
const protocol = isHttpProtocol ? rawProtocol : 'https';

// Include port if it exists (regardless of protocol)
const host = port
  ? `${protocol}://${resource}:${port}`
  : `${protocol}://${resource}`;
```

---

## Overall Assessment

**Status**: ⚠️ Needs Changes

The PR addresses a real bug and the core fix is sound, but it needs adjustments for:
- **Completeness**: Missing Gitea service
- **Consistency**: Three different implementations
- **Correctness**: Port handling edge cases
- **Testing**: Additional test coverage needed

The implementation is 80% there but needs these refinements before merging to ensure all Git providers work correctly and the codebase remains maintainable.

---

## Additional Questions for Author

1. Have you tested this with actual GitHub Enterprise/GitLab/Bitbucket Server instances?
2. Are there any known consumers of `RepoMetadata.host` outside the git-service package?
3. Should this fix be backported to previous versions?
4. Does the plugin API expose `RepoMetadata` to external plugins?

**Comment by yanpzhan** [comment]
Checked on cluster launched against the latest pr code.
On 'Import from Git' page, In Git Repo URL in put 'http://gitea.local:3000/user/repo'. Filtered 'local' in 'Network' tab.
Could see the api request is using http schema with port: 'http://gitea.local:3000/api/v1/repos/user/repo'
/verified by yanpzhan

**Comment by rhamilto** [comment]
Solid work, @krishagarwal278!  
/lgtm

**Comment by krishagarwal278** [comment]
thanks so much! @rhamilto ❤️
